benches/vdf/rsa_group.rs

```diff
@@ -0,0 +1,135 @@
+#[macro_use]
+extern crate criterion;
+
+use criterion::Criterion;
+use rug::{integer::Order, Integer};
+use sha2::{Digest, Sha256};
+
+/// algo_2 from the paper
+fn verify(modulus: &Integer, g: &Integer, t: u64, y: &Integer, pi: &Integer) -> bool {
+    let modulus = modulus.clone();
+
+    let l = hash_to_prime(&modulus, &[&g, &y]);
+
+    let r = Integer::from(2).pow_mod(&Integer::from(t), &l).unwrap();
+    let pi_l = pi.clone().pow_mod(&l, &modulus).unwrap();
+    let g_r = g.clone().pow_mod(&r, &modulus).unwrap();
+    let pi_l_g_r = pi_l * g_r;
+
+    Integer::from(pi_l_g_r.div_rem_floor(modulus.clone()).1) == y.clone()
+}
+
+/// algo_3 from the paper
+fn eval(modulus: &Integer, g: &Integer, t: u64) -> (Integer, Integer) {
+    let modulus = modulus.clone();
+
+    // y <- (g^2)^t
+    let mut y = g.clone();
+    for _ in 0..t {
+        y = y.clone() * y.clone();
+        y = y.div_rem_floor(modulus.clone()).1;
+    }
+
+    let l = hash_to_prime(&modulus, &[&g, &y]);
+
+    // algo_4 from the paper, long division
+    // TODO: consider algo_5 instead
+    let mut b: Integer;
+    let mut r = Integer::from(1);
+    let mut r2: Integer;
+    let two = Integer::from(2);
+    let mut pi = Integer::from(1);
+
+    for _ in 0..t {
+        r2 = r.clone() * two.clone();
+        b = r2.clone().div_rem_floor(l.clone()).0;
+        r = r2.clone().div_rem_floor(l.clone()).1;
+        let pi_2 = pi.clone().pow_mod(&two, &modulus).unwrap();
+        let g_b = g.clone().pow_mod(&b, &modulus).unwrap();
+        pi = pi_2 * g_b;
+    }
+    pi = Integer::from(pi.div_rem_floor(modulus.clone()).1);
+    (y, pi)
+}
+
+/// int(H("residue"||x)) mod N
+fn h_g(modulus: &Integer, seed: &Integer) -> Integer {
+    let modulus = modulus.clone();
+    let mut hasher = Sha256::new();
+    hasher.update("residue".as_bytes());
+    hasher.update(seed.to_digits::<u8>(Order::Lsf));
+    let hashed = Integer::from_digits(&hasher.finalize(), Order::Lsf);
+
+    // inverse, to get enough security bits
```

Comment:
I double checked and we can actually cannot do the inverse trick as it do not provide a random element in the group.
Instead you should concat results of nine sha256 and take the result modulo N

Reply:
Should I run recursive hashing only on the imtermediate hashing result itself, then at the end use them to reconstruct the randomness?

Or should I hash->construct->hash repeatedly?

The latter seems can offer more entropy?

i.e., should I 
```
let mut temp = h_g_inner(seed);
let mut result = temp.clone();
let mut ent = HASH_ENT;
while ent < GROUP_ENT {
    temp = h_g_inner(&temp);
    result = (result << HASH_ENT) + temp.clone();
    ent += HASH_ENT;
}
result.div_rem_floor(modulus.clone()).1
``` 
or
```
let mut result = h_g_inner(seed);
let mut ent = HASH_ENT;
while ent < GROUP_ENT {
    result = h_g_inner(&result);
    result = (result.clone() << HASH_ENT) + result.clone();
    ent += HASH_ENT;
}
result.div_rem_floor(modulus.clone()).1
```

Reply:
I also come up with another idea.

We can actually

sha256("part1"||seed) || sha256("part1"||seed) || ... || sha256("part1"||seed) 

i.e., we keep using the same seed for input, but introduces "partXXX" to provide different randomness.

We then concat then all.

Reply:
I think all of the methods above works and we should take the fastest